### DISCUSSION TITLE
STM32G474: Performance issues (?) and freq() issues.

### DISCUSSION THREAD

**Discussion by nspsck** [micropython/micropython#13015] [STM32 / Pyboard]
I recently recieved my STM32G474CE boards made by WeAct Studio and started testing Micropython on it. Soon I have noticed that the `freq()` function does not work as expected and the performance from this thing is way worse than the F4 series.

Here are some results in the REPL using the `freq()` function:
```python
>>> pyb.freq()
(170000000, 170000000, 42500000, 85000000) # 170MHz
>>> pyb.freq(96000000) # 96MHz
>>> pyb.freq()
(8000000, 8000000, 2000000, 4000000) # 8MHz
>>> pyb.freq(170000000) # 170MHz
>>> pyb.freq()
(16400000, 16400000, 4100000, 8200000) # 16.4MHz
```
It seems like the the clock is divided by 10 for some reasons. Also, according to the datasheet in [the section about pll characteristics](https://www.st.com/resource/en/datasheet/stm32g474cb.pdf), it seems the MCU needs different voltage for different clock frequencies which I do not see covered anywhere in the `powerctl.c`. ([This section.](https://github.com/micropython/micropython/blob/6117aa686f5faac6692f0def91be8e40a696aca3/ports/stm32/powerctrl.c#L384-L707)) Am not so sure if that would cause any problems.

I have also tried to modify the `pllvalues.py` by adding a component to the `mcu_table` list declared in [line 34](https://github.com/micropython/micropython/blob/6117aa686f5faac6692f0def91be8e40a696aca3/ports/stm32/boards/pllvalues.py#L34):
```python
    "stm32g4": MCU(
        range_sysclk=range(2, 170 + 1),
        range_m=range(1, 8 + 1), # Can go 16, but in the Manual it changes PLLM to PLLSYSM
        range_n=range(8, 127 + 1),
        range_p=range(2, 31 + 1),
        range_q=range(2, 8 + 1, 2),
        range_vco_in=range(3, 16 + 1), # 2.66 - 16
        range_vco_out=range(96, 344 + 1),
    ),
```
But this results to a Fatal error: `can't change freq`. Could this also just be my chip?

For the performance part, I have used 2 simple tests:
```python
import time

# Int
def test(p):
    r = range(p)
    c = 0
    for i in r:
        c += 1
    return c

# Float
# def test(p):
#     r = range(p)
#     c = 0.0001
#     for i in r:
#         c += c
#         c = c / 2
#     return c * 4 / p

def main():
    ms = time.ticks_ms
    diff = time.ticks_diff
    
    while True:
        s = ms()
        print(test(50000))
        d = diff(ms(), s)
        print("Time used %dms" % d)
        
main()
```

Here are some results running my "benchmark":
```python
# STM32G474CE @170MHz
# Float
MPY: sync filesystems
MPY: soft reboot
8e-09
Time used 1074ms
8e-09
Time used 1076ms
8e-09
Time used 1076ms
# Int
MPY: sync filesystems
MPY: soft reboot
50000
Time used 292ms
50000
Time used 292ms
50000
Time used 293ms
50000

# STM32F411CE @120MHz
# Float
MPY: sync filesystems
MPY: soft reboot
8e-09
Time used 795ms
8e-09
Time used 796ms
8e-09
Time used 797ms
# Int
MPY: sync filesystems
MPY: soft reboot
50000
Time used 167ms
50000
Time used 167ms
50000
Time used 167ms
50000
```
The results is really odd for me, since at least the Integer operation should not differ that much...

Any help and suggestions are greatly appreciated!

**Comment by ajt2018** [comment]
Could it be related to the fact that the WeAct board costs £0.40 from AliExpress yet the STM32G431CBU6 on it's own would set you back £2.6 if you bought 1000's of them from digikey ?

**Reply by nspsck** [reply]
> the WeAct board costs £0.40 from AliExpress

That is not totally true, it's a deal only for the first time you buy it and only limited to 1 piece. It will cost you roughly £4.78 each after that. Especially the `freq()` part, I had some success modifying the `pllvalues.py`. And performance wise, I think it's ralated to the CCM RAM based on [this topic](https://community.st.com/t5/stm32-mcus-products/project-migration-from-stm32f405rg-to-stm32g473re/td-p/253565). However, I haven't tested it myself any further because I couldn't get the quad spi flash working and that is a deal breaker for me for this chip.

**Reply by ajt2018** [reply]
Sorry, I was being a bit sarcastic there! 
We use the G474 with HAL without any problem but I don't think I'd notice a performance drop anyway - we reduce the clocks to save on power usage anyway.
I'll remember the CCM tip though if I need more speed in the future :)

**Reply by robert-hh** [reply]
The Nucleo G474 board has a 24MHz crystal, while the WEACT board has a 8MHz crystal. If you use the WEACT board with the Nucleo settings, then that will be one reason why the speed will be lower.

**Reply by nspsck** [reply]
Hmm, should it matter after I set up the PLLs? Or could it be the `pllvalues.py`? By default, it generates pll table for the G4 series using F4 configurations. But I have no idea what those values are and how to configure them correctly.  And by the above mentioned "some success" only means it boot up after I modified those values in `pllvalues.py`. Calling `freq()` still messes up the frequencies. Hence I have not tested the performance myself after changing it.

**Reply by robert-hh** [reply]
Did you call pllvalues.py with a value of 8 for the HSE`?

**Reply by nspsck** [reply]
By `pllvalues.py` I am refering to [this auxiliary script](https://github.com/micropython/micropython/blob/master/ports/stm32/boards/pllvalues.py). Which I can not call directly. It directly reads the .h files with HSE informations. What I have tried was:

in `mpconfigboard.h`:
```c
// The board has an 8MHz HSE, the following gives 170MHz CPU speed
#define MICROPY_HW_CLK_PLLM         (2)
#define MICROPY_HW_CLK_PLLN         (85)
#define MICROPY_HW_CLK_PLLP         (2)
#define MICROPY_HW_CLK_PLLQ         (8)
#define MICROPY_HW_CLK_PLLR         (2)
```
in `stm32g4xx_hal_conf.h`:
```c
// Oscillator values in Hz
#define HSE_VALUE               (8000000)
#define LSE_VALUE               (32768)
#define EXTERNAL_CLOCK_VALUE    (24000)
```
As far as I know, the `EXTERNAL_CLOCK_VALUE`  is only used for the i2s interface and should not affect the main cpu frequency. But I could be wrong on this one, since I have **not** done a deep research on this.

**Reply by robert-hh** [reply]
That should be sufficient.

**Reply by ajt2018** [reply]
I'd highly recommend downloading and running ST CubeMX and the Clock Configuration tab.
You will be able to pick the STM32G4 and configure the external clock, PLL multipliers, dividers and bus speeds, then generate some code which will show you the values for PLL's etc. I find it far easier to visualise all this inside CubeMX and just pull out the settings I need for my own code.
(See image of the clock configuration in this StackOverflow question: https://stackoverflow.com/questions/49481275/generating-system-clock-configuration-for-stm32f103c8-with-cubemx)
I would have thought that the Quad SPI didn't work because of clock setup (it's always related to clock settings when peripherals don't work!)

**Reply by nspsck** [reply]
Thanks for the recommendation! Actually, that is the software I am already using. And yes, the quad spi issue is likely caused by the clock settings, but if I continue to use micropython, unless I verify all the registers using the stm module and the user manual, there is (to me) no otherway to verify the actual clock signal, because `freq()` is kinda broken on the G4 series right now. Unless your experience on this particular chip is different, which implies that I must have done something wrong, badly.

**Reply by davefes** [reply]
@nspsck, I need a STM32 with more GPIO than the STM32F411.  Did you manage to get this variant working properly?

**Reply by nspsck** [reply]
@davefes hi, sadly not. But have a look at: https://de.aliexpress.com/item/1005006051651136.html?spm=a2g0n.productlist.0.0.2eff31fffELZ4A&browser_id=bbdc4688ff7e4fa6ba7bf004319ef9fb&aff_platform=msite&m_page_id=hhdhvrtgcablurld192fc38c48c11080024e5781a3&gclid=&pdp_npi=4%40dis%21CHF%214.77%211.54%21%21%2138.47%2112.39%21%40211b80e117308084317374384efa9b%2112000035503740772%21sea%21CH%210%21ABX&algo_pvid=ce2ee2ab-73ee-4872-9918-e73c750397c9

The F412RET6. Double the RAM, almost same price, but you can not solder a spi flash directly on the board.

**Reply by davefes** [reply]
Downloaded the datasheet, looking at the package and trying to count the pins in the ad I'd guess it is a LQFP64 package.  That is 50 GPIO.

I would like 80 GPIO.  

Would the STM32G474CE board from WeACT work OK for GPIO? I want to control 25 stepper motor controller boards, so only GPIO toggling is required.

Thanks.

**Reply by nspsck** [reply]
That C in STM32G474CE means that this soc has 48 pins. 

C - 48 pins
R - 64 pins
V - 100 pins
Z - 144 pins

You might want something like a STM32F407VE or STM32F407VG

B - 128KB flash
E - 512KB flash
G - 1MB flash

Or the STM32H723ZG/VG, STM32H743VG. STM32H750VB if you know how to get the qspi flash working.

**Reply by davefes** [reply]
Thanks for the clarification.  I had considered the STM32H723VGT6 from WeAct.  The actual program size is relatively small, so if on-board flash works then I might be OK.
https://micropython.org/download/NUCLEO_H723ZG/

1MB of flash, which is good enough for me.
